File: LeetCode/LeetCode-CN-JAVA/2020六月每日一题/1028. 从先序遍历还原二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode recoverFromPreorder(String S) {
        Stack<TreeNode> path = new Stack<TreeNode>();
        int pos = 0;
        while(pos < S.length()) {
            int level = 0;
            while(S.charAt(pos) == '-') { // 计算出下个层数
                level++;
                pos++;
            }
            int value = 0; // 计算出下个数字
            while(pos < S.length() && Character.isDigit(S.charAt(pos))) {
                value = value*10 + Integer.valueOf(S.charAt(pos) - '0');
                pos++;
            }
            // System.out.println(value);
            TreeNode node = new TreeNode(value);
            // level == size() 的时候说明，level正好是size()的下一层，因为size()表示已经存在了多少层，由于是从0层开始，所以加入size=2，说明已经存在了0层和1层，此时level=2说明下个节点正好应该在下一层。
            if(level == path.size()) { // 直接相等时说明当前节点应该插入到上个节点的左边
                if(!path.empty()) { // 只有相等的情况下，可能会出现level = 0 , size = 0 的初始情况，其余情况下栈不会为空
                    path.peek().left = node;
                }
            } else {
                while(level != path.size()) { // 不等时说明当前节点应该是之前路径上某个节点的右节点，所以需要回溯到之前的层次
                    path.pop();
                }
                path.peek().right = node;
            }
            path.push(node);
        }
        while(path.size() > 1) {
            path.pop();
        }
        return path.peek();
    }
}

File: LeetCode/LeetCode-CN-CPP/2020六月每日一题/1028. 从先序遍历还原二叉树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // 假设输入格式合法
    TreeNode* recoverFromPreorder(string S) {
        if(S.size() == 0) {
            return NULL;
        }
        int num_start_index = 0; // 当前数字的起始位置
        int num_end_index = S.find("-") ; // 当前数字的最后位置 + 1
        int cur_level = 0; // 记录当前的层次
        TreeNode* root = new TreeNode(stoi(S.substr(num_start_index, num_end_index)));
        TreeNode* p = root;
        stack<TreeNode*> stk;
        stk.push(p);
        // 即便结束时 stk也不为空
        while(!stk.empty()) {
            TreeNode* node = NULL;
            // 判断下个节点是不是下一层的
            size_t next_num_start = S.find_first_not_of("-", num_end_index); // 找到下一个非“-”的位置
            if(next_num_start == string::npos) { // 说明已经到了字符串末尾, 真正的结束条件
                return root;
            }
            size_t next_num_end = S.find("-", next_num_start); // 找到下一个“-”的位置 ， 也是下一个数字位置+1
            if(next_num_end == string::npos) { // 说明已经是最后一个数字
                next_num_end = S.size(); // 记录为字符串的长度
            }
            int next_level = next_num_start  - num_end_index; // 得到下面是第几层的
            int next_num = stoi(S.substr(next_num_start, next_num_end)); // 提取出下一个数字
            //cout << "cur_level:" << cur_level << " , next_level:" << next_level;
            //cout << "cur_num:" << p->val << " , next_num:" << next_num <<  " ,stknum:" << stk.size() <<endl;
            if(next_level > cur_level) { // 如果是当前层下面的层的话就添加到当前节点的左节点
                node = new TreeNode(next_num);
                p->left = node;
                p = node;
                //cout << "》链接左节点，链接后的值为：" << p->val << endl;
                stk.push(node);
                cur_level = next_level;
                num_start_index = next_num_start;
                num_end_index = next_num_end;
                continue;
            }

            if(next_level == cur_level) { // 如果下一层等于当前层，说明是同级别的
                stk.pop(); // 当前层节点出栈
                p = stk.top(); // 获取上一层的节点
                node = new TreeNode(next_num); 
                p->right = node;
                p = node;
                //cout << "=链接右节点，链接后的值为：" << p->val << endl;
                stk.push(node);
                cur_level = next_level;
                num_start_index = next_num_start;
                num_end_index = next_num_end;
                continue;
            }
            if(next_level < cur_level) {
                stk.pop();
                p = stk.top();
                //cout << "《此时的值为：" << p->val << endl;
                cur_level--;
            }
        }
        return root;
    }
};